### ISSUE TITLE
SceneManager.OnLoadEnd does not guarantee spawn order on clients

### ISSUE THREAD

**Issue by mk0a1a** [FirstGearGames/FishNet#98]
The order when spawning new NetworkObjects during SceneManager.OnLoadEnd is not guaranteed for clients. For example, after a scene load with the ReplaceOption.All option, the NetworkObject spawned on OnLoadEnd is **spawned first** before the scene game objects. My scripts break because it depends on the managers present in the scene. Note that the server spawns these in the correct order so there are no issues there.

I'm aware of OnClientLoadedStartScenes as this is used to spawn players during the initial scene load. But as far as I know, this only gets called one time when clients connects to the server the first time.

TLDR: I need to guarantee the order of spawns. with similar functionality as SceneManager.OnClientLoadedStartScenes but for every scene load.

**Comment by FirstGearGames** [comment]
I suspect you aren't using the network behavior callbacks. Please ask for help in regards to this in the discord.

**Comment by mk0a1a** [comment]
Hi, I'm actually using proper NetworkBehaviour callbacks. In fact, I do all network object initializations on OnStartXXX. The bug I'm reporting on this issue is that the OnStartXXX callbacks are getting **called first before** the scene objects are spawned. With the event being named OnLoadEnd on the SceneManager, you'd expect things to be done **after**. (Note: issue is only for clients, the order on the server is correct)

I reported it here first instead of asking on discord because I thought this was a pretty clear cut bug (the order of operations between the server and client are clearly desynced), not something I need help implementing.

I logged the order of events below on the server:
![onstartnetwork server order](https://user-images.githubusercontent.com/44934887/187060570-e7aa385c-dccf-43c6-8fee-cdafca5bff69.png)

Here is the same logs but on the client side:
![onstartnetwork client order](https://user-images.githubusercontent.com/44934887/187060603-1ed634e4-a30b-4c85-9983-8ae54aa0a956.png)

The "OnStartNetwork: ..." logs are done using a simple static counter that increments each time they are called and they are getting called in OnStartNetwork (I tried it on OnStartClient and it is the same). You can even see on the client side logs that it can't find the gameobject "Scene GameObject" which is a scene NetworkObject with a NetworkObserver conditioned to be on scene.

Anyways, I've already written a workaround for this bug using RPCs instead of SceneManager events so I can accept this issue being closed. I've read all pages from the documentation and watched all tutorial videos and I don't think I see this being mentioned there. But could we at least have something written in the documentation something like "Don't spawn things on SceneManager.OnLoadEnd as spawn order is not guaranteed on the client side"?

**Comment by FirstGearGames** [comment]
Based on your response this does not seem like a bug. Scene objects are sent to the client after presence change start and before presence change end. You're spawning the object before either of those events.

But also, because of how dictionaries work objects still may have their spawn messages sent in orders different from when they were added to the dictionary. None the less, they will all arrive the same tick and all objects will have sync data initialized before network callbacks. The order of callbacks cannot be promised due to reasons just mentioned.

**Comment by mk0a1a** [comment]
Thanks for the clarification! So to be clear, the OnClientPresenceChangeStart and OnClientPresenceChangeEnd are the proper events to do spawning. I've already written a robust system using RPCs so I won't be changing it but I'll keep this in mind in the future. Thanks again!

**Comment by FirstGearGames** [comment]
Not necessarily. They can be used for a variety of reasons and you can
spawn before them. If you'd like more information on this please ping me in
the server.
